File: python/tracing/google-genai/01_generate_content.py
from __future__ import annotations

from respan import workflow

from _shared import (
    example_attributes,
    make_client,
    make_custom_identifier,
    make_respan,
    model_name,
    print_result,
    workflow_name,
)

EXAMPLE_NAME = "generate-content"


@workflow(name=workflow_name(EXAMPLE_NAME))
def _generate_content_workflow(client) -> str:
    response = client.models.generate_content(
        model=model_name(),
        contents="Reply with one concise sentence about observability for Gemini apps.",
    )
    return response.text or ""


def run_generate_content() -> None:
    respan = make_respan(EXAMPLE_NAME)
    client = make_client()
    custom_identifier = make_custom_identifier(EXAMPLE_NAME)
    text = ""

    try:
        with example_attributes(EXAMPLE_NAME, custom_identifier):
            print(f"custom_identifier={custom_identifier}", flush=True)
            print(f"workflow_name={workflow_name(EXAMPLE_NAME)}", flush=True)
            text = _generate_content_workflow(client)
    finally:
        respan.flush()
        respan.shutdown()

    print_result(EXAMPLE_NAME, custom_identifier, text)


if __name__ == "__main__":
    run_generate_content()

File: typescript/tracing/vertexai/01_generate_content.ts
import {
  createVertexExampleRuntime,
  flushAndShutdown,
  logExampleResult,
  responseFromResult,
  runWithExampleTrace,
  textFromResponse,
  WEATHER_TOOL,
} from "./_shared.js";

const workflowName = "TypeScript VertexAI Generate Content Example";

export async function generateContentExample(): Promise<void> {
  const { mode, model, respan } = await createVertexExampleRuntime(
    "typescript-vertexai-generate-content-example",
    {
      systemInstruction: "Answer in one concise sentence.",
      tools: [WEATHER_TOOL],
      generationConfig: {
        maxOutputTokens: 128,
        temperature: 0.2,
        topP: 0.9,
        topK: 32,
      },
    },
  );

  try {
    const result = await runWithExampleTrace(respan, workflowName, async () =>
      await model.generateContent({
        contents: [
          {
            role: "user",
            parts: [{ text: "Write one sentence about why trace observability matters." }],
          },
        ],
      }),
    );
    const response = await responseFromResult(result);
    logExampleResult(workflowName, {
      mode,
      text: textFromResponse(response),
      usage: response?.usageMetadata,
    });
  } finally {
    await flushAndShutdown(respan);
  }
}

await generateContentExample();
